src/lib/operators/join_hash/multi_predicate_join_evaluator.hpp

```diff
@@ -11,134 +11,113 @@
-    std::vector<ColumnID> left_col_ids;
-    std::vector<ColumnID> right_col_ids;
-
-    for (const auto& pred : _join_predicates) {
-      left_col_ids.push_back(pred.column_id_pair.first);
-      right_col_ids.push_back(pred.column_id_pair.second);
-
-      resolve_data_type(_left.column_data_type(pred.column_id_pair.first), [&](auto type) {
-        using ColumnDataType = typename decltype(type)::type;
-
-        switch (pred.predicate_condition) {
-          case PredicateCondition::Equals:
-            _comparators.emplace_back(std::make_unique<EqVoidComparator<ColumnDataType>>());
-            break;
-          case PredicateCondition::GreaterThan:
-            _comparators.emplace_back(std::make_unique<GtVoidComparator<ColumnDataType>>());
-            break;
-          case PredicateCondition::GreaterThanEquals:
-            _comparators.emplace_back(std::make_unique<GtEqVoidComparator<ColumnDataType>>());
-            break;
-          case PredicateCondition::LessThan:
-            _comparators.emplace_back(std::make_unique<LtVoidComparator<ColumnDataType>>());
-            break;
-          case PredicateCondition::LessThanEquals:
-            _comparators.emplace_back(std::make_unique<LtEqVoidComparator<ColumnDataType>>());
-            break;
-          case PredicateCondition::NotEquals:
-            _comparators.emplace_back(std::make_unique<NEqVoidComparator<ColumnDataType>>());
-            break;
-          default:
-            throw std::runtime_error("Predicate condition not supported!");
-        }
+  MultiPredicateJoinEvaluator(const Table& left, const Table& right,
+                              const std::vector<JoinPredicate>& join_predicates) {
+    for (const auto& pred : join_predicates) {
+      resolve_data_type(left.column_data_type(pred.column_id_pair.first), [&](auto left_type) {
+        using LeftColumnDataType = typename decltype(left_type)::type;
+        resolve_data_type(right.column_data_type(pred.column_id_pair.second), [&](auto right_type) {
+          using RightColumnDataType = typename decltype(right_type)::type;
+
+          std::vector<std::unique_ptr<AbstractSegmentAccessor<LeftColumnDataType>>> left_accessors;
+          std::vector<std::unique_ptr<AbstractSegmentAccessor<RightColumnDataType>>> right_accessors;
+
+          left_accessors = _create_accessors<LeftColumnDataType>(left, pred.column_id_pair.first);
+          right_accessors = _create_accessors<RightColumnDataType>(right, pred.column_id_pair.second);
+
+          // todo: copied from column_vs_column_table_scan_impl
+          constexpr auto LEFT_IS_STRING_COLUMN = (std::is_same<LeftColumnDataType, std::string>{});
+          constexpr auto RIGHT_IS_STRING_COLUMN = (std::is_same<RightColumnDataType, std::string>{});
+
+          constexpr auto NEITHER_IS_STRING_COLUMN = !LEFT_IS_STRING_COLUMN && !RIGHT_IS_STRING_COLUMN;
+          constexpr auto BOTH_ARE_STRING_COLUMNS = LEFT_IS_STRING_COLUMN && RIGHT_IS_STRING_COLUMN;
```

Comment:
I thought (maybe I'm wrong) joining only works for columns of the same type? If so, you could save a bit of code here.

Reply:
Negative. We can join all types (well, except for string combinations)